Copenhagen Business School

MSc in Business Administration and Data Science

Natural Language Processing and Text Analytics

Mandatory Assignment 1

Students:

- Joel Berglund - 167681

- William Keller Boll - 167326

- Eirik Lile Vågen - 167300

- Carl Tegnér - 167302

# QUESTION 1
The following two sentences are ambiguous – that is, they each have two (or more) different readings.
For each sentence, explain the different readings:
1. The girl attacked the boy with the book.
2. We decided to leave on Saturday.

Are the following sentences ambiguous? If the sentence is ambiguous, explain the different readings. If not, explain why not.

3. I saw a man with a briefcase.
4. I saw the planet with a telescope.

Try the above sentences (3 and 4) in this parse tree generator:
https://huggingface.co/spaces/nanom/syntactic_tree
Compare the dependency structures for both of them. Explain the differences in the link to the
preposition “with” in 3 vs. 4.

Sentence 1: The girl attacked the boy with the book.

Sentence 2: We decided to leave on Saturday.

The two sentences above are ambiguous. They can be understood in more than just one single way. If we start off with the first example, “with the book” is the part that leaves the understanding of the sentence ambiguous. We would argue that there are two main interpretations of this sentence. The first being that the girl attacks the boy using a book, physically using the book to hit the boy. The second interpretation would be that the girl attacks the boy, and the book being in his possession is what is used to describe the boy.

The second sentence is ambiguous around the on Saturday part. The first interpretation would be that they (We) had, at an earlier point in time before the sentence was uttered, decided to leave the on the following Saturday from when the sentence was uttered. The second interpretation could be that the decision to leave was decided the last Saturday from when the sentence was uttered.

The following two sentences both use with as a preposition, but in different ways. After this section, the sentences and their dependency structures can be seen. The sentences are as follows “I saw a man with a briefcase.” and “I saw the planet with a telescope.”. In the first sentence, with is used to describe the man, like in the second interpretation of sentence 1, where the book is used to describe the boy. We would argue that it is virtually unambiguous, as everyone would understand that a briefcase is used to describe the man, and not used as a tool to see the man.

The same goes for the last sentence. In this sentence, with is used as a preposition to describe how the subject saw a planet. The telescope was a tool that facilitated and allowed the subject to see the planet. One could argue that one, in very special cases, could interpret the sentence as a planet being identified as the one with a telescope. But virtually everyone would interpret the sentence as the subject using a telescope to see the planet.


Sentence 3: I saw a man with a briefcase.

saw: (Root) (VBD)

├── I: 1 (PRP)

├── man: 1 (NN)

│   ├── a: 2 (DT)

│   └── with: 2 (IN)

│       └── briefcase: 3 (NN)

│           └── a: 4 (DT)

└── .: 1 (.)

Sentence 4: I saw the planet with a telescope.

saw: (Root) (VBD)

├── I: 1 (PRP)

├── planet: 1 (NN)

│   └── the: 2 (DT)

├── with: 1 (IN)

│   └── telescope: 2 (NN)

│       └── a: 3 (DT)

└── .: 1 (.)



# QUESTION 2 - [Linguistic Analysis of a Text Corpus Using spaCy]



In [ ]:
import spacy
import nltk
from nltk import word_tokenize
from nltk.corpus import stopwords
from nltk.probability import FreqDist
from nltk import pos_tag
from prettytable import PrettyTable
import spacy.cli
from collections import Counter

from google.colab import drive
drive.mount('/content/drive')

# Download the spaCy English model, if not already installed
spacy.cli.download("en_core_web_sm")

# Load spaCy model
nlp = spacy.load("en_core_web_sm")

nltk.download('stopwords')
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

1. Prepare the Corpus: Pre-process the corpus data by removing stopwords, removing extra
spaces, converting all text to lowercase, or apply any other text normalization techniques that
you think is relevant.

2. POS Tagging and NER: Apply POS tagging to the entire corpus to analyze the distribution of
different parts of speech. Use NER to identify and categorize named entities within the text
such as name, date, locations etc.

In [ ]:
# Define your custom stop words
custom_stop_words = ['!', "'", "''", "'d", "'em", "'ll", "'m", "'re", "'s", "'t", "'ve", ',', '--', '.', '112._', '1920',
                    '26._', '45._', '5._', ':', ';', '?', '[', ']', '_71_', '_could_', '_had_', '_has_', '_him_', '_his_',
                    "_i'd_", '_our_', '_page', '_their_', '_was_', '``', "n't"]

# Add custom stopwords to NLTK's list
stop_words = set(stopwords.words('english')) | set(custom_stop_words)



file_path_drive = '/content/drive/MyDrive/CBS/Övningsuppgifter T2/NLP/sample.txt'  # Path for Google Colab
file_path_local = 'sample.txt'  # Local file path for Jupyter Notebook

try:
    # Try to open the file assuming it's in Google Drive
    with open(file_path_drive, 'r', encoding='utf-8') as file:
        text = file.read().lower()
except FileNotFoundError:
    # If the file is not found, try to open it from a local path
    try:
        with open(file_path_local, 'r', encoding='utf-8') as file:
            text = file.read().lower()
    except FileNotFoundError:
        print("File not found. Please check your file path.")


tokens = word_tokenize(text)

# Filter out stop words and non-alphabetic tokens
filtered_tokens = [word for word in tokens if word not in stop_words and word.isalpha()]


# Generate frequency distribution
freq_dist = FreqDist(filtered_tokens)
# Sort by frequency (already in descending order)
sorted_by_freq = sorted(freq_dist.items(), key=lambda item: item[1], reverse=True)
print(sorted_by_freq)

# Apply POS tagging to filtered tokens
pos_tags = pos_tag([word for word, freq in sorted_by_freq])

# Create a dictionary for quick POS tag lookup
pos_tags_dict = {word: tag for word, tag in pos_tags}

# Process the text with spaCy for NER, create a dictionary for quick NER lookup
doc = nlp(' '.join(filtered_tokens))
ner_dict = {ent.text: ent.label_ for ent in doc.ents}

# Initialize PrettyTable
table = PrettyTable(['Word', 'Freq', 'POS', 'NER'])

# Initialize a list for words with a named entity tag
words_with_ne = []

# Fill the table with data and prepare words_with_ne list
for word, freq in sorted_by_freq:
    pos_tag = pos_tags_dict.get(word, 'N/A')
    ne_tag = ner_dict.get(word, 'None')  # Default to 'None' if word is not a named entity
    table.add_row([word, freq, pos_tag, ne_tag])

    # If the word has a NE tag other than 'None', add it to words_with_ne list
    if ne_tag != 'None':
        words_with_ne.append((word, ne_tag))

print(table)

3. Analysis: Determine the most common POS tags and discuss what this reveals about the structure of the text.

In [ ]:
# Count the frequency of each POS tag
pos_tags_freq = Counter([tag for word, tag in pos_tags])

# Count the frequency of each NER tag
ner_tags_freq = Counter(ner_dict.values())

# Print the frequencies of each POS tag in descending order
print("POS Tag Frequencies (Descending Order):")
for tag, freq in pos_tags_freq.most_common():
    print(f"{tag}: {freq}")

print("\nNER Tag Frequencies (Descending Order):")
for tag, freq in ner_tags_freq.most_common():
    print(f"{tag}: {freq}")

POS Tag Frequencies (Descending Order):
NN: 373
JJ: 279
VBD: 127
RB: 123
NNS: 100
VBG: 91
VBN: 78
VBP: 42
VB: 25
VBZ: 24
JJS: 23
IN: 20
JJR: 12
MD: 7
RBR: 5
NNP: 5
CD: 3
DT: 3
CC: 2
RP: 1
WP$: 1
PRP: 1
WP: 1
FW: 1
WRB: 1
WDT: 1

NER Tag Frequencies (Descending Order):
PERSON: 32
TIME: 10
DATE: 9
ORG: 7
CARDINAL: 4
ORDINAL: 2
NORP: 1
QUANTITY: 1


The high frequency of nouns (NN: 373, NNS: 100, NNP: 5) and adjectives (JJ: 279, JJR: 12, JJS: 23) indicates a descriptive text, possibly narrative, focused on detailing specific subjects, objects, and their characteristics. The variety of verb forms, including the past tense (VBD: 127), gerunds/participles (VBG: 91), past participles (VBN: 78), present tense (VBP: 42, VBZ: 24), and the base form (VB: 25), suggests a mix of actions and states, possibly recounting events or describing ongoing situations.

4. Prepare a short report: Summarize your findings in a brief report of 1-2 page, including any
interesting patterns or insights gained from the POS and NER analysis.

From our POS and NER analysis, we have discovered the following. Our NER analysis revealed that the text places a strong emphasis on personal entities (PERSON: 32), temporal references (TIME: 10), (DATE: 9), and organizations (ORG: 7). This suggests that the text narrates human-centric stories and involves elements of a social or institutional context, for instance, a government institution or a social organization, where the characters are engaged. An example of these NER tags could be, "Michael Johnsson (PERSON) embarked on a journey in the early spring (DATE) of 1923 (DATE)."

However, our NER analysis also showed that occurrences of cardinal (CARDINAL: 4), ordinal (ORDINAL: 2), and quantity (QUANTITY: 1) are less frequent in the text. These NER tags are more commonly found, for example, in recipes or perhaps in sports commentary, where they might discuss, "on his third attempt, he broke the world record."

Our POS analysis reveals a high use of nouns (NN: 373, NNS: 100, NNP: 5) and adjectives (JJ: 279, JJR: 12, JJS: 23), indicating the text is descriptive or narrative, with detailed descriptions of states or entities. Additionally, the analysis reveals that the occurrences of verbs in various forms (VBD: 127, VBG: 91, VBN: 78, VBP: 42, VB: 25, VBZ: 24) indicate a dynamic narrative with a mix of actions. An example of this could be, "She whispered (VBD) softly, her fingers tracing (VBG) the rough texture of the old photograph (NN)."

Furthermore, the POS analysis also suggests that the frequency of prepositions (IN: 20), conjunctions (CC: 2), and determiners (DT: 3) indicate a more straightforward narrative style. For example, a lower frequency might lead to sentences that lack temporal context, for instance: "He walked the forest." versus "He walked through the forest," or "She works night." versus "She works at night." However, this conclusion is not definitive since the text is not extensive, and the number of prepositions present isn't substantial.

To conclude our analysis reveals that the text is a story of interpersonal relationships portrayed through descriptions, all inside a societal or institutional setting. The narrative paints its scenes more with words than with numbers, prioritizing the exploration of personal tales above numerical facts. It is, in short, a human-centered storytelling that is rich in action.